### PR TITLE
Convert RABX columns to JSONB

### LINKED ISSUE #3216: Switch from RABX to JSON
It didn't when this started, but now PostgreSQL has JSON support and if we switched our custom (protocol buffer-like) RABX columns to JSON that would be more standard and allow better querying.

### PR DESCRIPTION
**Pull request by dracos**
This has been a bit tricky, hopefully it all makes sense. Fixes #3216

In terms of the whole diff for this PR, that does what you would expect:
* the columns change from text or bytea to jsonb
* the code changes to make JSON queries and inflate/deflate as JSON (only a couple of places beyond the automatic);
* One change to "order" because RABX always stores stringed numbers as integers, whereas JSON does not, which caused an update-when-no-update-necessary issue
* Similarly with `_eq_column_values`, the data comes out of the database ordered by key-length then alphabetical, whereas going to the database is only alphabetical, so we do a little to/from dance just so the "has this column changed" code continues to work. Think that's okay.
* No indexes, so this probably won't make any difference performance-wise, but shouldn't be worse, and they can be added afterwards to hopefully speed up the querying of these columns.
* A script that runs as part of the database migration to backfill all existing data. If you don't mind the site being down for the duration, a normal deploy/update running update-schema should be okay, copy all data across and Just Work.

The main issue is the migration of data, as that could take a long time and we do not want to take down the .com site for the duration. So this PR is two separate commits, which can be deployed separately. The first adds the new columns and runs backfill, the second drops the old column and renames. But still, the backfill could take a long time. So if you want to have the minimal downtime possible, I think the plan is:
1. ~~Run migration 82 manually (except the NOT VALID constraint line, as that would then fail!), to add new columns.~~
2. ~~Deploy first commit of this PR (backfill script will not run as column already exists). This now means any update to a RABX column updates the JSON column too.~~
3. ~~Add the NOT VALID constraint from migration 82, as any updates will be setting the column now the code is deployed.~~
4. ~~Run script to backfill any existing entries.~~
5. ~~Pause, all data should now be in sync between RABX and JSON columns and be being kept in sync.~~
6. ~~Run migration 83 manually to sort out the NOT NULL thing on Token, just in case that takes a while.~~
7. ~~At time site is not busy, turn off front end servers~~
8. ~~TAKE BACKUP OF DB, JUST IN CASE I MESSED SOMETHING UP (c. 15 minutes).~~
9. Deploy second commit of this PR on one live server; there will be some small downtime while columns are dropped and renamed (migration 84).
10. Deploy other front end servers.
11. Reenable disabled front end servers in balancer
12. Have an ice cream.

Commands for each of the above:

7. Run from the management server. Removes all hosts from load balancers, then stops each.
```
for PORT in 6091 6092 ; do
    for SERVER in swfms001hexa swfms001sova swfms002sova ; do
        sudo varnishadm -S /etc/varnish/secret -T localhost:${PORT} backend.set_health boot.www_fixmystreet_com_${SERVER}_srv_mysociety_org sick
    done
done

for SERVER in swfms001hexa swfms001sova swfms002sova ; do
    sudo mysociety vhost stop --server $SERVER www.fixmystreet.com
done
```
8. On the database server: `pg_dump --schema=public -Fc bci > bci.sql.dump`
9. Run driectly on an app server, assume swfms002sov: `mysociety vhost www.fixmystreet.com`
10. Second deploys that don't need migration can be run centrally from management systems:
```
mysociety vhost --server swfms001hexa www.fixmystreet.com
mysociety vhost --server swfms001sova www.fixmystreet.com
```
11.
```
for PORT in 6091 6092 ; do
    for SERVER in swfms001hexa swfms001sova swfms002sova ; do
        sudo varnishadm -S /etc/varnish/secret -T localhost:${PORT} backend.set_health boot.www_fixmystreet_com_${SERVER}_srv_mysociety_org auto
    done
done
```

This is all yak shaving so that I can add geolocation_used as a key to extra on Problem because I did not think it needed its own column. Let's hope no-one disagrees with me on that! ;)

### CODE REVIEW

**Review by struan — CHANGES_REQUESTED** [review]
I did an upgrade and then a downgrade and the downgrading seems to break some geocode data. I've got a copy of some of the live reports and it broke on this one:

https://www.fixmystreet.com/report/1043482

with

`bad netstring trailer character 'n'`

It's fine if I convert the JSON from the upgrade back to RABX by hand so I think it's the process of putting it in the database? If I look at the raw bytes in the database they are different (original 1st, converted back from JSON 2nd):
 \x41313a372c5432343a61757468656e7469636174696f6e526573756c74436f64652c5431363a56616c696443726564656e7469616c732c5431323a6272616e644c6f676f5572692c5435363a687474703a2f2f6465762e7669727475616c65617274682e6e65742f4272616e64696e672f6c6f676f5f706f77657265645f62792e706e672c54393a636f707972696768742c543234373a436f7079726967687420c382c2a92032303137204d6963726f736f667420616e642069747320737570706c696572732e20416c6c207269676874732072657365727665642e2054686973204150492063616e6e6f7420626520616363657373656420616e642074686520636f6e74656e7420616e6420616e7920726573756c7473206d6179206e6f7420626520757365642c20726570726f6475636564206f72207472616e736d697474656420696e20616e79206d616e6e657220776974686f75742065787072657373207772697474656e207065726d697373696f6e2066726f6d204d6963726f736f667420436f72706f726174696f6e2e2c5431323a7265736f75726365536574732c4c313a312c41313a322c5431343a657374696d61746564546f74616c2c49313a312c54393a7265736f75726365732c4c313a312c41313a392c54363a5f5f747970652c5436313a4c6f636174696f6e3a687474703a2f2f736368656d61732e6d6963726f736f66742e636f6d2f7365617263682f6c6f63616c2f77732f726573742f76312c54373a616464726573732c41313a372c5431313a616464726573734c696e652c5431323a506f7474657273204c616e652c5431333a61646d696e44697374726963742c54373a456e676c616e642c5431343a61646d696e4469737472696374322c54353a48657274732c5431333a636f756e747279526567696f6e2c5431343a556e69746564204b696e67646f6d2c5431363a666f726d6174746564416464726573732c5433303a506f7474657273204c616e652c2053746576656e6167652c2053473120322c54383a6c6f63616c6974792c54393a53746576656e6167652c5431303a706f7374616c436f64652c54353a53473120322c54343a62626f782c4c313a342c5231363a35312e383933373438323832343239332c5231383a2d302e3232343037373636333130383532392c5231363a35312e393031343733373137353730372c5231383a2d302e3230373338363333363839313437312c5431303a636f6e666964656e63652c54363a4d656469756d2c5431303a656e74697479547970652c54373a416464726573732c5431333a67656f636f6465506f696e74732c4c313a312c41313a342c5431373a63616c63756c6174696f6e4d6574686f642c5431333a496e746572706f6c6174696f6e2c5431313a636f6f7264696e617465732c4c313a322c52393a35312e3839373631312c52393a2d302e3231353733322c54343a747970652c54353a506f696e742c5431303a757361676554797065732c4c313a322c54373a446973706c61792c54353a526f7574652c5431303a6d61746368436f6465732c4c313a312c54343a476f6f642c54343a6e616d652c5433303a506f7474657273204c616e652c2053746576656e6167652c2053473120322c54353a706f696e742c41313a322c5431313a636f6f7264696e617465732c4c313a322c52393a35312e3839373631312c52393a2d302e3231353733322c54343a747970652c54353a506f696e742c5431303a737461747573436f64652c49333a3230302c5431373a7374617475734465736372697074696f6e2c54323a4f4b2c54373a747261636549642c5436373a35646566633135663630383034626336613063376134373062383936613363667c444234303036303233357c372e372e302e307c4442345343483031303036313732372c

vs 

\x41313a372c54373a747261636549642c5436373a35646566633135663630383034626336613063376134373062383936613363667c444234303036303233357c372e372e302e307c4442345343483031303036313732372c5431373a7374617475734465736372697074696f6e2c54323a4f4b2c5431323a6272616e644c6f676f5572692c5435363a687474703a2f2f6465762e7669727475616c65617274682e6e65742f4272616e64696e672f6c6f676f5f706f77657265645f62792e706e672c5431323a7265736f75726365536574732c4c313a312c41313a322c54393a7265736f75726365732c4c313a312c41313a392c5431333a67656f636f6465506f696e74732c4c313a312c41313a342c5431373a63616c63756c6174696f6e4d6574686f642c5431333a496e746572706f6c6174696f6e2c5431313a636f6f7264696e617465732c4c313a322c52393a35312e3839373631312c52393a2d302e3231353733322c5431303a757361676554797065732c4c313a322c54373a446973706c61792c54353a526f7574652c54343a747970652c54353a506f696e742c5431303a656e74697479547970652c54373a416464726573732c54343a6e616d652c5433303a506f7474657273204c616e652c2053746576656e6167652c2053473120322c54353a706f696e742c41313a322c5431313a636f6f7264696e617465732c4c313a322c52393a35312e3839373631312c52393a2d302e3231353733322c54343a747970652c54353a506f696e742c54343a62626f782c4c313a342c5231363a35312e383933373438323832343239332c5231383a2d302e3232343037373636333130383532392c5231363a35312e393031343733373137353730372c5231383a2d302e3230373338363333363839313437312c5431303a6d61746368436f6465732c4c313a312c54343a476f6f642c54373a616464726573732c41313a372c5431333a636f756e747279526567696f6e2c5431343a556e69746564204b696e67646f6d2c5431343a61646d696e4469737472696374322c54353a48657274732c5431333a61646d696e44697374726963742c54373a456e676c616e642c5431313a616464726573734c696e652c5431323a506f7474657273204c616e652c5431363a666f726d6174746564416464726573732c5433303a506f7474657273204c616e652c2053746576656e6167652c2053473120322c5431303a706f7374616c436f64652c54353a53473120322c54383a6c6f63616c6974792c54393a53746576656e6167652c5431303a636f6e666964656e63652c54363a4d656469756d2c54363a5f5f747970652c5436313a4c6f636174696f6e3a687474703a2f2f736368656d61732e6d6963726f736f66742e636f6d2f7365617263682f6c6f63616c2f77732f726573742f76312c5431343a657374696d61746564546f74616c2c49313a312c54393a636f707972696768742c543234373a436f7079726967687420c2a92032303137204d6963726f736f667420616e642069747320737570706c696572732e20416c6c207269676874732072657365727665642e2054686973204150492063616e6e6f7420626520616363657373656420616e642074686520636f6e74656e7420616e6420616e7920726573756c7473206d6179206e6f7420626520757365642c20726570726f6475636564206f72207472616e736d697474656420696e20616e79206d616e6e657220776974686f75742065787072657373207772697474656e207065726d697373696f6e2066726f6d204d6963726f736f667420436f72706f726174696f6e2e2c5431303a737461747573436f64652c49333a3230302c5432343a61757468656e7469636174696f6e526573756c74436f64652c5431363a56616c696443726564656e7469616c732c

in case that helps.

This happens every time I run it.

FWIW, the upgrade seems to be fine.

**Comment by dracos** [comment]
Yes, I thought saving bytea wouldn't encode, but it does at the low level the migration script is written in (as I'm not using bind_param). Hopefully fixed in 6b7e68d4daacdab45e1e8c31ee0714fa64553805, along with similar fix in extra for JSON (doesn't need to utf8), and catching geocode rows without extra.

**Review by sagepe — COMMENTED** [review]
There's quite a lot happening here, and I'm not vastly familiar with all the internal details so I'm going to leave the more in-depth code review to Struan and Dave. I'm also going to assume that any fixups will be squashed into the most appropriate places.

The general procedure you outline seems fine; as I understand it step (1) can be run manually outside a deploy as this will simply add the new columns that nothing will use them until the first commit is deployed in step (2) and this deploy can happen as normal (one node can start adding to both columns while the other continues with just RABX) - also the backfill script won't trigger as update-schema will see us already at 75, rather than moving up to it. 

Steps (3)-(5) seem uncontroversial - I assume we can backfill in the background without problem, I don't think I see a reason why not. Ditto with step (6).

At step (7), I'd suggest we remove the server in question from the load-balancer, then stop the vhost to ensure no local jobs are running. Then at step (10) we'd want to re-enable this in the load-balancer once the deploy had completed. There are example commands for doing this on the wiki: https://wiki.mysociety.org/wiki/Web_Server_Stack#Controlling_and_Checking_Back-Ends

At step (8) for belt-and-braces I'd probably add the commands needed to the plan - but this is simply enough to do.

I'd also want to expand slightly on what exactly needs to be done at step (9) - have an implementation plan that someone other than the author would be comfortable following.

Finally, I'd advise running through this process in full in staging if that's not already been completed.

**Comment by dracos** [comment]
Ooh, that gave me a good idea, I've split the NOT NULL bit into a separate schema change, which should make the steps clearer and simpler, will update them now.

**Review by struan — APPROVED** [review]
Few small things but this looks good.

Converted to json and back to rabx and then back to json fine locally.

**Comment by sagepe** [comment]
One slight tweak: when removing from Load Balancers, best to remove from both just in case (FMS is actually on Bittern right now, I think):
```
for balancer in bittern starling ; do
  varnishadm -S /etc/varnish/secret -T ${balancer}:6082 backend.set_health boot.www_fixmystreet_com_crow sick
done
```

File: perllib/FixMyStreet/DB/Result/AdminLog.pm [hunk 1]
```diff
@@ -33,9 +34,8 @@ __PACKAGE__->add_columns(
   "whenedited",
   {
     data_type     => "timestamp",
-    default_value => \"current_timestamp",
+    default_value => \"CURRENT_TIMESTAMP",
```
[the remark this answers is not in the document]
Comment (dracos) [reply perllib/FixMyStreet/DB/Result/AdminLog.pm:37]: Auto-generation, yeah, different db version perhaps? Not sure.

File: perllib/Open311/PostServiceRequestUpdates.pm [hunk 2]
```diff
@@ -94,7 +94,10 @@ sub construct_query {
         'me.external_id' => undef,
         'me.state' => 'confirmed',
         'me.confirmed' => { '!=' => undef },
-        'me.extra' => [ undef, { -not_like => '%cobrand_skipped_sending%' } ],
+        -or => [
+            'me.extra' => undef,
+            -not => { 'me.extra' => { "\\\?" => 'cobrand_skipped_sending' } }
```
[the remark this answers is not in the document]
Comment (dracos) [reply perllib/Open311/PostServiceRequestUpdates.pm:99]: I should have done a single quote. So "?" returns an "unbound placeholder", as does "\?" as in double quoted string that becomes a "?" too. So three in order for it be passed through as '\?' but single quote would have been clearer, will swap!

**Review by davea — APPROVED** [review]
Looks good to me. Worked OK on my local DB, including the downgrade. 👍

**Review by sagepe — APPROVED** [review]

**Comment by davea** [comment]
Just to note that this would make FD-1901 possible/easier.

**Comment by chrismytton** [comment]
Note that https://github.com/mysociety/fixmystreet/pull/3494 has been merged the `roles` table also has an `extra` column that uses RABX.

**Pull request merged**